src/lib/apply-shim.html

```diff
@@ -0,0 +1,197 @@
+<!--
+@license
+Copyright (c) 2014 The Polymer Project Authors. All rights reserved.
+This code may only be used under the BSD style license found at http://polymer.github.io/LICENSE.txt
+The complete set of authors may be found at http://polymer.github.io/AUTHORS.txt
+The complete set of contributors may be found at http://polymer.github.io/CONTRIBUTORS.txt
+Code distributed by Google as part of the polymer project is also
+subject to an additional IP rights grant found at http://polymer.github.io/PATENTS.txt
+-->
+<link rel="import" href="style-util.html">
+<script>
+Polymer.ApplyShim = (function(){
+  'use strict';
+
+  var styleUtil = Polymer.StyleUtil;
+
+  var MIXIN_MATCH = styleUtil.rx.MIXIN_MATCH;
+  var VAR_ASSIGN = styleUtil.rx.VAR_ASSIGN;
+  var VAR_MATCH = styleUtil.rx.VAR_MATCH;
+  var APPLY_NAME_CLEAN = /;\s*/m;
+
+  // map of mixin to property names
+  // --foo: {border: 2px} -> (--foo, ['border'])
+  var mixinMap = {};
+
+  function mapSet(name, prop) {
+    name = name.trim();
+    mixinMap[name] = prop;
+  }
+
+  function mapGet(name) {
+    name = name.trim();
+    return mixinMap[name];
+  }
+
+  // return properties that are not used by the new definition of a mixin
+  // diffProperties([a, b], [b]) -> [a]
+  function diffProperties(oldSet, newSet) {
+    var diff = [];
+    if (oldSet && oldSet.length) {
+      for (var i = 0, old; i < oldSet.length; i++) {
+        old = oldSet[i];
+        if (newSet.indexOf(old) === -1) {
+          diff.push(old);
+        }
+      }
+    }
+    return diff;
+  }
+
+  // produce variables in place of a mixin definition
+  function defineMixinProperties(mixinName, propNames, propMap, oldProps) {
+    var out = [];
+    var i, p, v;
+    // set variables defined by current mixin
+    for (i = 0; i < propNames.length; i++) {
+      p = propNames[i];
+      v = propMap[p];
+      out.push(mixinName + '-' + p + ': ' + v);
+    }
+    // reset variables that are overriden by the current mixin
+    var resetProps = diffProperties(oldProps, propNames);
+    for (i = 0; i < resetProps.length; i++) {
+      p = resetProps[i];
+      out.push(mixinName + '-' + p + ': initial');
+    }
+    return out.join(';\n') + ';';
```

Comment:
Let's consider omiting the `\n` here. I think it doesn't enhance readability much due to indenting issues.